tests/test_circuit_graph.py

```diff
@@ -260,11 +280,71 @@ def test_observable_nodes_and_operation_nodes(self, circuit, monkeypatch):
         values."""
         mock_graph = MagicMock()
         mock_graph.nodes.values.return_value = [
-            {"return_type": None, "idx": 0}, {"return_type": MagicMock(), "idx": 1}]
+            {"return_type": None, "idx": 0},
+            {"return_type": MagicMock(), "idx": 1},
+        ]
         monkeypatch.setattr(circuit, "_graph", mock_graph)
         assert circuit.observable_nodes == [mock_graph.nodes.values()[1]]
         assert circuit.operation_nodes == [mock_graph.nodes.values()[0]]
 
     def test_operations(self, circuit, queue):
         """Test that the `operations` property returns the list of operations in the circuit."""
         assert circuit.operations == queue
+
+    def test_get_names(self, circuit):
+        result = circuit.get_names(range(9))
+        expected = ["RX", "RY", "RZ", "CNOT", "Hadamard", "CNOT", "PauliX", "PauliX", "Hermitian"]
+        assert result == expected
+
+    def test_op_indices(self, circuit):
+        """Test that for the given circuit, this method will fetch the correct operation indices for
+        a given wire"""
+        op_indices_for_wire_0 = [0, 3, 5, 7]
+        op_indices_for_wire_1 = [1, 3, 6, 8]
+        op_indices_for_wire_2 = [2, 4, 5, 8]
+
+        assert circuit.get_op_indices_for_wire(0) == op_indices_for_wire_0
+        assert circuit.get_op_indices_for_wire(1) == op_indices_for_wire_1
+        assert circuit.get_op_indices_for_wire(2) == op_indices_for_wire_2
+
+    def test_layers(self, parameterized_circuit):
+        """A test of a simple circuit with 3 layers and 6 parameters"""
+        dev = qml.device("default.qubit", wires=3)
+        circuit = qml.QNode(parameterized_circuit, dev)
+        circuit.construct((0.1, 0.2, 0.3, 0.4, 0.5, 0.6))
```

Comment:
Just curious, how come you used `construct` here?

Reply:
if I recall, the circuit graph is only created during construct, so must be called for the `QNode.circuit` attribute to exist.

Reply:
The `CircuitGraph` instance gets generated and set in `QNode.construct` -- since we don't need to evaluate anything or do any further processing, I manually called `construct` with a bunch of parameters, which associates the parameters with the operations and creates the DAG, etc... We should probably  refactor `construct` and break it down into more specific methods.